### PR TITLE
WIP: All siles now supports pathlib paths; tests added

### PR DESCRIPTION
**Pull request by jonaslb**
At least they support it as far as the tests show. I guess that this should be tested on older Pythons (3.5, 2.7) to ensure it works there as well, but it should work, as far as I can see.

### CODE REVIEW

**Comment by jonaslb** [comment]
So it turns out py2.7 compat is not too great still, I must have forgotten it in a few places. Will look into it later.

**Comment by zerothi** [comment]
Once done, this will fix #112

**Comment by jonaslb** [comment]
This should be ready for merging when py 2.7 support is dropped, as I will not go through the effort to support that. I tested it on 3.6 and 3.7, but it should also work on 3.5.

**Comment by zerothi** [comment]
Great work. I'll review this :)

File: sisl/io/tbtrans/tbt.py [hunk 1]
```diff
@@ -2742,7 +2743,7 @@ def copy_attr(f, t):
                 v[:] = dvg[:]
 
         # Update the source attribute to signal the originating file
-        self.setncattr('source', 'k-average of: ' + tbt._file)
+        self.setncattr('source', 'k-average of: ' + str(tbt._file))
```
Comment (zerothi) [inline sisl/io/tbtrans/tbt.py]: casting to `str` should be implicit, I don't think it is needed?

File: sisl/io/tests/test_object.py [hunk 2]
```diff
@@ -3,6 +3,12 @@
 import pytest
 
 import numpy as np
+import sys
+import os
+try:
+    from pathlib import Path
```
Comment (zerothi) [inline sisl/io/tests/test_object.py]: delete `try...except` since pathlib is a requirement

File: sisl/io/tests/test_object.py [hunk 3]
```diff
@@ -24,7 +30,9 @@ def _my_intersect(a, b):
 
 
 def _fnames(base, variants):
-    return [base + '.' + v if len(v) > 0 else base for v in variants]
+    files = [base + '.' + v if len(v) > 0 else base for v in variants]
```
Comment (zerothi) [inline sisl/io/tests/test_object.py]: `base` is not a `Path`? Should it?

Comment (zerothi) [reply sisl/io/tests/test_object.py]: It seems so...

File: sisl/io/tests/test_object.py [hunk 4]
```diff
@@ -56,6 +64,24 @@ class TestObject(object):
     def test_siesta_sources(self):
         pytest.importorskip("sisl.io.siesta._siesta")
 
+    def test_direct_path_instantiation(self):
+        fp = Path("/tmp/shouldnotexist.1234567")
```
Comment (zerothi) [inline sisl/io/tests/test_object.py]: use `Path("tmp") / Path("...")`, since we may support windows in the not too far future. `osp.join` or `Path` intrinsic should be used

File: sisl/io/tests/test_object.py [hunk 5]
```diff
@@ -56,6 +64,24 @@ class TestObject(object):
...
+        if fp.exists():
+            os.remove(str(fp))
+        for Sile in get_siles():
+            sile = Sile(fp, _open=False)
+            assert isinstance(sile, Sile)
+            assert not fp.exists()
+
+    def test_direct_string_instantiation(self):
+        fp = "/tmp/shouldnotexist.1234567"
```
Comment (zerothi) [inline sisl/io/tests/test_object.py]: same as above

File: sisl/io/tests/test_table.py [hunk 6]
```diff
@@ -19,8 +19,8 @@ def test_tbl1(sisl_tmp):
     io0.write_data(dat0, dat1)
     io1.write_data((dat0, dat1))
 
-    F0 = open(io0.file).readlines()
-    F1 = open(io1.file).readlines()
+    F0 = open(str(io0.file)).readlines()
```
Comment (zerothi) [inline sisl/io/tests/test_table.py]: use `Path.open`

File: sisl/io/tests/test_table.py [hunk 7]
```diff
@@ -33,8 +33,8 @@ def test_tbl2(sisl_tmp):
...
```
Comment (zerothi) [inline sisl/io/tests/test_table.py]: same

File: sisl/io/wannier90/seedname.py [hunk 8]
```diff
@@ -53,14 +54,14 @@ class winSileWannier90(SileWannier90):
     def _setup(self, *args, **kwargs):
         """ Setup `winSileWannier90` after initialization """
         self._comment = ['!', '#']
-        self._seed = self.file.replace('.win', '')
+        self._seed = str(self.file).replace('.win', '')
```
Comment (zerothi) [inline sisl/io/wannier90/seedname.py]: perhaps `_seed` should also be a `Path`, then `_seed.with_suffix('.win')` would work

File: sisl/io/wannier90/seedname.py [hunk 9]
```diff
@@ -53,14 +54,14 @@ class winSileWannier90(SileWannier90):
...
 
     def _set_file(self, suffix=None):
         """ Update readed file """
         if suffix is None:
-            self._file = self._seed + '.win'
+            self._file = Path(self._seed + '.win')
         else:
-            self._file = self._seed + suffix
+            self._file = Path(self._seed + suffix)
```
Comment (zerothi) [inline sisl/io/wannier90/seedname.py]: see above

File: sisl/io/xsf.py [hunk 10]
```diff
@@ -378,10 +378,8 @@ class axsfSile(Sile):
     """
 
     def _setup(self, *args, **kwargs):
+        kwargs["_md_steps"] = kwargs.get("_md_steps", 1)
```
Comment (zerothi) [inline sisl/io/xsf.py]: I don't think this belongs in this PR, this PR is already quite extensive and this should go in another PR.

File: sisl/io/sile.py [hunk 11]
```diff
@@ -699,7 +698,7 @@ def __init__(self, filename, mode='r', lvl=0, access=1, *args, **kwargs):
             # The CDF file can easily open the file
         if kwargs.pop('_open', True):
             _import_netCDF4()
-            self.__dict__['fh'] = _netCDF4.Dataset(self.file, self._mode,
+            self.__dict__['fh'] = _netCDF4.Dataset(str(self.file), self._mode,
```
[the remark this answers is not in the document]
Comment (jonaslb) [reply sisl/io/sile.py]: Yes, but I looked it up: Py3.5 was released 2015/9 and pathlib support was merged (no idea about released) in netcdf in 2015/11 (see https://github.com/Unidata/netcdf4-python/pull/490 ). It is possible that if someone out there is using Py3.5, that they might also use an older version of netcdf4.

File: sisl/io/tbtrans/tbt.py [hunk 12]
```diff
@@ -2537,7 +2537,8 @@ def __call__(self, parser, ns, value, option_string=None):
                 else:
                     ns._tbt.write_tbtav(value)
         p.add_argument('--tbt-av', action=AVOut, nargs='?', default=None,
-                       help='Create "{0}" with the k-averaged quantities of this file.'.format(self.file.replace('TBT.nc', 'TBT.AV.nc')))
+                       help='Create "{0}" with the k-averaged quantities of this file.'.format(
+                        str(self.file).replace('TBT.nc', 'TBT.AV.nc')))
```
[the remark this answers is not in the document]
Comment (jonaslb) [reply sisl/io/tbtrans/tbt.py]: Unfortunately it's not so easy with multiple suffixes, there isn't a `.with_suffixes` method :( It could be `self.file.with_name(self.file.name.replace('TBT.nc', 'TBT.AV.nc'))` but that's hardly more elegant...

File: sisl/io/tbtrans/tbt.py (hunk 1 above)
Comment (jonaslb) [reply sisl/io/tbtrans/tbt.py]: I can change it to a string with `format`, but `+` only works if both are strings.

File: sisl/io/wannier90/seedname.py (hunk 8 above)
Comment (jonaslb) [reply sisl/io/wannier90/seedname.py]: I don't know if it every happens, but can't the seed contain a dot? In that case we'd be overwriting the last part.

File: sisl/io/sile.py [hunk 13]
```diff
@@ -513,10 +511,11 @@ def __init__(self, filename, mode='r', comment=None, *args, **kwargs):
         self._base_setup(*args, **kwargs)
 
     def _open(self):
```
[the remark this answers is not in the document]
Comment (jonaslb) [reply sisl/io/sile.py]: It would be simpler but it's not a common practice, as far as I know. So it might be easier to write the code, but it'd be harder to read.

File: sisl/io/xsf.py (hunk 10 above)
Comment (jonaslb) [reply sisl/io/xsf.py]: It is necessary because otherwise, the later access to `_md_steps` causes it to try to open the file for reading, and then it fails the tests (where it's given a nonexistent file).

Comment (jonaslb) [reply sisl/io/xsf.py]: Actually come to think of it the precise line you commented on is not necessary. Just the deleted ones below were necessary to delete. Does the `_md_steps` even matter?

Comment (jonaslb) [reply sisl/io/xsf.py]: I think it should inherit from xsfSile instead of Sile and then it would matter to pass the `steps` parameter instead of `_md_steps`. But like you said that might be for another PR. Right now I guess it should just pass the tests.

File: sisl/io/tests/test_object.py [hunk 14]
```diff
@@ -1,14 +1,12 @@
 from __future__ import print_function, division
 
+import tempfile
```
Comment (zerothi) [inline sisl/io/tests/test_object.py]: Instead of using `tempfile` you should add an argument `sisl_tmp` to the test which requires a temporary file.
That would ensure you have a temporary file based on the same test-directory. See below

File: sisl/io/tests/test_object.py [hunk 15]
```diff
@@ -65,22 +63,24 @@ def test_siesta_sources(self):
...
 
     def test_direct_path_instantiation(self):
-        fp = Path("/tmp/shouldnotexist.1234567")
-        if fp.exists():
-            os.remove(str(fp))
-        for Sile in get_siles():
-            sile = Sile(fp, _open=False)
-            assert isinstance(sile, Sile)
-            assert not fp.exists()
+        with tempfile.TemporaryDirectory() as dir:
```
Comment (zerothi) [inline sisl/io/tests/test_object.py]: Change to:

```python
    def test_direct_path_instantiation(self, sisl_tmp):
fp = sisl_tmp.file('sisl', 'shouldnotexist.1234657')
```
which will create a temporary file for you (and automatically delete it once the test is done.

For details, see `sisl/conftest.py`, I would suggest you to add `pathlib` to that file as well. :)

Thanks!
￼

File: sisl/io/sile.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (zerothi) [reply sisl/io/sile.py]: yeah, probably that is correct. It is just annoying to call a function all the time... Lets keep it like this for now.

File: sisl/io/wannier90/seedname.py (hunk 8 above)
Comment (zerothi) [reply sisl/io/wannier90/seedname.py]: Doing 
```python
a = pl.Path('soaethuoase.')
a.with_suffix('.win')
PosixPath('soaethuoase..win')
```
yields the correct thing. :)

File: sisl/io/xsf.py (hunk 10 above)
Comment (zerothi) [reply sisl/io/xsf.py]: Agreed, I have fixed the inheritance in `axsf`. So if you rebase it should be fixed.
Weird I haven't had errors before?

**Comment by zerothi** [comment]
This has now been merged (I did a rebase to check conflicts step-by-step)
